### ISSUE TITLE
Poor performance despite low network/cpu usage

### ISSUE THREAD

**Issue by kaysond** [borgbackup/borg#4843]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question

## System information. For client/server mode post info for both machines.
Client: 8c Xeon E5-2695v2 VM
Server: Raspberry Pi 3B

#### Your borg version (borg -V).
1.1.10 on both

#### Operating system (distribution) and version.
Client: Ubuntu 18.04.3
Server: Raspbian 9.9

#### Hardware / network configuration, and filesystems used.
Client: ext4 on vmfs on Crucial SSD
Server: ext4 on USB drive (mechanical)

Network: WAN, limited by 20Mbps upload on client

#### How much data is handled by borg?
420GB de-duped

#### Full borg commandline that lead to the problem (leave away excludes and passwords)
`borg create --verbose --compression zstd,10 "::$DESTARCHIVE-{now}" $SOURCEPATH`
Where the command vars and other env vars (repo, ssh key, etc) are set elsewhere in a script

## Describe the problem you're observing.
Backing up a ~5GB folder takes at least a couple of hours. Raw throughput of the network connection says it should take about half an hour.

Network usage on the client is so low it rarely registers on `nethogs`. Occasionally I see it jump up to ~300KBps.

On the client, borg is using about 6% cpu on the client, and very little memory.

On the server, I also see nearly no cpu usage, and about 100MB RAM usage.

I'd expect to see the process either limited by CPU on the server due to the compression, or by the network, but its neither... Throughput on the CRUD benchmark also seems pretty horrible, despite no compression. I'm wondering if its some issue with indexing/de-dupe but I figured that would spike CPU.

CRUD Benchmarks:
```
C-Z-BIG           8.80 MB/s (10 * 100.00 MB all-zero files: 113.59s)
R-Z-BIG           6.59 MB/s (10 * 100.00 MB all-zero files: 151.75s)
U-Z-BIG           8.59 MB/s (10 * 100.00 MB all-zero files: 116.40s)
D-Z-BIG          10.27 MB/s (10 * 100.00 MB all-zero files: 97.38s)
C-R-BIG           1.74 MB/s (10 * 100.00 MB random files: 574.12s)
R-R-BIG           0.94 MB/s (10 * 100.00 MB random files: 1058.74s)
U-R-BIG           8.74 MB/s (10 * 100.00 MB random files: 114.38s)
D-R-BIG           6.92 MB/s (10 * 100.00 MB random files: 144.61s)
C-Z-MEDIUM        9.55 MB/s (1000 * 1.00 MB all-zero files: 104.74s)
R-Z-MEDIUM        0.69 MB/s (1000 * 1.00 MB all-zero files: 1456.17s)
U-Z-MEDIUM        8.74 MB/s (1000 * 1.00 MB all-zero files: 114.43s)
D-Z-MEDIUM       10.11 MB/s (1000 * 1.00 MB all-zero files: 98.91s)
C-R-MEDIUM        1.74 MB/s (1000 * 1.00 MB random files: 574.64s)
R-R-MEDIUM        0.87 MB/s (1000 * 1.00 MB random files: 1155.78s)
U-R-MEDIUM        8.83 MB/s (1000 * 1.00 MB random files: 113.19s)
D-R-MEDIUM        6.47 MB/s (1000 * 1.00 MB random files: 154.51s)
C-Z-SMALL         0.95 MB/s (10000 * 10.00 kB all-zero files: 105.70s)
R-Z-SMALL         1.24 MB/s (10000 * 10.00 kB all-zero files: 80.65s)
U-Z-SMALL         0.86 MB/s (10000 * 10.00 kB all-zero files: 116.15s)
D-Z-SMALL         1.04 MB/s (10000 * 10.00 kB all-zero files: 95.94s)
C-R-SMALL         0.68 MB/s (10000 * 10.00 kB random files: 146.10s)
R-R-SMALL         1.39 MB/s (10000 * 10.00 kB random files: 71.96s)
U-R-SMALL         0.87 MB/s (10000 * 10.00 kB random files: 114.32s)
D-R-SMALL         0.91 MB/s (10000 * 10.00 kB random files: 109.68s)
```

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.
Yes, happens consistently

#### Include any warning/errors/backtraces from the system logs
No warnings or errors

**Comment by ThomasWaldmann** [comment]
There can be many reasons for this, including:
- borg is currently single threaded, so many cores won't help (fast cores help, to a certain degree)
- borg does a sequential mix of cpu intensive stuff (e.g. computing hashes) and I/O intensive stuff (reading/writing, from/to disk or network)
- filesystem and network performance, througput and latency
- many small files are more effort than fewer bigger ones
- the raspi design is not for performance (lots of stuff going over usb multiple times, little and not expandable RAM, slow cores)
- too high/slow compression (not sure for zstd,10 - zstd in general is nice stuff and relatively fast, check if level 10 is slowing it down due to higher compression)

The initial backup is relatively slow in general as it has to process a lot of stuff. But that is only once, future backups are way faster as it has only the changes to process and can deal with unchanged files very quickly.

**Comment by fantasya-pbem** [comment]
You could also try lz4 compression which is the fastest algorithm.

**Comment by kaysond** [comment]
@ThomasWaldmann Are my CRUD benchmark throughputs normal? Because <1MBps is not really usable for 100's of GB... I'm better off just rsyng'ing the whole thing...

Also I tried re-running the exact command, and it took even longer! 2hrs for the first backup, and 2.5hrs for the subsequent one. I'll try the experiment again with less compression.

One thing I did think about is that I'm also doing a prune after every upload because I only want to keep the last one, but figured it would be faster if most things don't change. Could this be part of the issue? Is there a better way of doing this?

**Comment by ThomasWaldmann** [comment]
I ran the benchmark on my laptop, via wifi to a local machine and it was 13MB/s for the C-R-BIG.

What's usable depends on the amount of changes per time period. I did daily backups over < 1Mbps for a long time and it worked well because the changes per day were small enough and borg made sure only to transmit the stuff that was really different. IIRC I did the first backup by transporting a disk to the remote, back then.

I perf problems are recent, it could also be that you have a problem in your setup somewhere:
- your internet connection being slower than it should be (or more loaded, so less is available for borg)
- your hardware having issues (e.g. if hard disks grow defects, they can be super slow at some defect sectors while giving still good performance on all others)

Only keeping the last backup isn't a thing one usually does. You have deduplication, so keeping multiple backups is cheap (not extremely many, but some - following a prune policy that fits your needs).

**Comment by kaysond** [comment]
Is it possible that multiple failed archive creations could somehow corrupt the repo?

This happened because I filled my remote disk...multiple times. (This is also why I'm keeping a single copy of each archive, for now. Storage upgrades coming).

I tried a different compression and observed some very strange behavior. Here's my backup log. Each backup is a `borg create` followed by a `borg prune -P <prefix> --keep-last 1`

First I ran a backup with `zstd,10`, which took `02:04`. The next day, I ran the same backup again, but this time it took `02:36`. Most of the files are git repos, and they haven't changed.

```
2019-11-18 17:18:53
Backing up /backups/rsnapshot/alpha.0/home/ to borg@backups.iamnub.com:/media/extdrive/OffsiteBackups/fserver::home
Backup completed in 02:04:27
2019-11-18 19:59:43
Backing up /backups/rsnapshot/alpha.0/fserver_root to borg@backups.iamnub.com:/media/extdrive/OffsiteBackups/fserver::fserver_root
Backup completed in 03:52:41
2019-11-19 08:47:01
Backing up /backups/rsnapshot/alpha.0/home/ to borg@backups.iamnub.com:/media/extdrive/OffsiteBackups/fserver::home
Backup completed in 02:36:31
```

Then I changed to `zstd,8`, and somehow the whole backup only took 4 minutes. I ran it again, and same thing.
```
2019-11-19 13:24:32
Backing up /backups/rsnapshot/alpha.0/home/ to borg@backups.iamnub.com:/media/extdrive/OffsiteBackups/fserver::home
Backup completed in 00:04:34
2019-11-19 14:09:46
Backing up /backups/rsnapshot/alpha.0/home/ to borg@backups.iamnub.com:/media/extdrive/OffsiteBackups/fserver::home
Archive created in 00:03:53
Backup completed in 00:05:54
```
(NB: I added an extra logging line before the last run to see the time difference between `create` and `prune`)

**Comment by kaysond** [comment]
So I just ran another backup. This one was about 80GB worth of data, compressed to about 70GB. De-duped to 25GB, which I assume is what actually went over the network. The backup took 13 hours...

I could've rsync'd the whole thing, uncompressed in about 4hrs with my upload connection.
```
Backing up /backups/aramdesk/ to borg@backups.iamnb.com:/media/extdrive/OffsiteBackups/fserver::aramdesk
Archive created in 13:05:26
Backup completed in 13:49:30
```

Is this just an issue with large repos? It seems I would be better off rsync-ing everything to an encrypted filesystem and using something like `rsnapshot` for "de-dup". Not exactly as efficient as borg, but its way faster.

**Comment by ThomasWaldmann** [comment]
If you want to do borg benchmarks for full (first) backups, you always need to start with a fresh repo or it won't be comparable.

For practical purposes, the long runtimes you'll see there are not much relevant though as the daily changes are not that much, so it'll finish much faster.

**Comment by kaysond** [comment]
> whether the files-cache works correctly [== ctime, size, inode number being stable for the borg defaults]).

Do you mean the inode of the file on the client?

Also, is the file cache on the client side or server? And is it one per repo? or shared between all?

Because I have a weekly backup schedule where a different location gets uploaded to borg each day. I wonder if that's ruining the cache?

**Comment by infectormp** [comment]
> Do you mean the inode of the file on the client?
Yes

You can read more about now it works here https://borgbackup.readthedocs.io/en/stable/internals.html

**Comment by kaysond** [comment]
Thanks for the help everyone.

I identified a couple of issues that seem to have improved overall throughput:
1. My backup drive was connected to USB2.0 and passed through to the VM. Performance here seems to have been the biggest limiting factor. Directly passing a PCIE USB3 controller caused the IOWAIT to drop from 95% to 30%, and CPU usage is now up to 70%, so it seems limited by the compression, which is much better.
2. Changing compression and changing inodes
The way I had my local incremental backups set up caused the inodes to change, invalidating the cache. I was also changing the compression to experiment, which I think also changes the chunks and probably also invalidated the cache.

Solved those two, and now my incremental backup took about 2:30hr for a 350GB dataset where about 200MB had changed. I'm guessing this could probably be sped up further if I turned off compression, at the cost of a slower initial backup.
